### PR TITLE
[feature](cold-hot) support remote storage

### LINKED ISSUE #8531: [Feature] Support remote storage for Cold-Hot data
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

## Background

To support the separation of hot and cold data, you need to specify the connection information of the remote storage when creating an OLAP table. 
For example, if the remote storage is the object storage S3, you need to specify the access key, secret key and other information. This information can be reused, and a Doris cluster may also be connected to multiple remote storages, so these connection information needs to be managed as remote storage. It is convenient for users to use, and it is not necessary to provide it every time a table is created, only the remote storage name can be provided. At the same time, when the remote connection information changes, there is no need to rebuild the table, only the corresponding remote storage information needs to be updated.

## Design

Need to increase the management function of remote storage, as follows

1. Support the creation of remote storage information
    1. At present, only S3 is supported, and more types may be supported in the future. Expansion design needs to be considered
2. Support remote storage information display
3. Support delete remote storage information

The remote storage information is stored in the catalog, and RemoteStorageManager is added to manage these meta information. When using it, it can be obtained from the manager according to the remote storage name.

When creating an OLAP table, add the `remote_storage` parameter and use the corresponding remote storage according to the remote_storage_name information. Support to modify the corresponding information through `ALTER TABLE`.

When the cold data storage medium `storage_cold_medium` is `s3`, the `remote_storage` information must be added. When the time reaches `storage_cooldown_time`, the cold data will be migrated to the remote S3.

### Specific syntax

```sql
-- create remote storage
alter system add remote storage "remote_s3"
properties
(
    "type" = "s3",
    "s3_endpoint" = "bj",
    "s3_region" = "bj",
    "s3_root_path" = "/path/to/root",
    "s3_access_key" = "bbb",
    "s3_secret_key" = "aaaa",
    "s3_max_connections" = "50",
    "s3_request_timeout_ms" = "3000",
    "s3_connection_timeout_ms" = "1000"
);

alter system add remote storage "remote_s3_1"
properties
(
    "type" = "s3",
    "s3_endpoint" = "bj",
    "s3_region" = "bj",
    "s3_root_path" = "/path/to/root",
    "s3_access_key" = "bbb",
    "s3_secret_key" = "aaaa",
    "s3_max_connections" = "50",
    "s3_request_timeout_ms" = "3000",
    "s3_connection_timeout_ms" = "1000"
);

-- view remote storage
show remote storage;

-- delete remote storage
alter system drop remote storage "remote_s3";

-- modify remote storage
alter system modify remote storage remote_s3 
properties
(
    "s3_max_connections" = "500"
);


-- Create OLAP table with remote storage information
-- When "storage_cold_medium" = "S3", "remote_storage" = "remote_s3" must exist
CREATE TABLE example_db.table_remote
(
    aa BIGINT,
    bb VARCHAR(64),
    cc VARCHAR(64),
    dd VARCHAR(64)
)
ENGINE=olap
DISTRIBUTED BY HASH (aa) BUCKETS 1
PROPERTIES(
    "replication_num" = "1",
    "storage_medium" = "HDD",
    "storage_cold_medium" = "S3",
    "remote_storage" = "remote_s3",
    "storage_cooldown_time" = "2022-04-01 20:24:00"
);

-- Modify the remote storage information of the OLAP table
alter table table_remote modify partition table_remote set ("storage_medium" = "HDD","storage_cold_medium" = "S3","remote_storage" = "remote_s3_1","storage_cooldown_time" = "2022-04-01 20:24:00");

```

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by qidaye**
# Proposed changes

Issue Number: close #8531 

1. Support add/drop/show/modify remote storage
2. Support create olap table with remote storage
3. Support modify data property of olap table

## Problem Summary:

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
4. Has unit tests been added: (Yes/No/No Need)
5. Has document been added or modified: (Yes/No/No Need)
6. Does it need to update dependencies: (Yes/No)
7. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/common/util/PropertyAnalyzer.java [hunk 1]
```diff
@@ -130,21 +139,48 @@ public static DataProperty analyzeDataProperty(Map<String, String> properties, D
...
+                }
+            } else if (!hasRemoteStorage && key.equalsIgnoreCase(PROPERTIES_REMOTE_STORAGE)) {
+                hasRemoteStorage = true;
+                remoteStorageName = value;
             }
         } // end for properties
 
-        if (!hasCooldown && !hasMedium) {
+        if (!hasCooldown && !hasMedium && !hasRemoteStorage) {
             return oldDataProperty;
```
Comment (pengxiangyu) [inline fe/fe-core/src/main/java/org/apache/doris/common/util/PropertyAnalyzer.java:156]: 如果是SSD->HDD，remote storage是空的。

File: fe/fe-core/src/main/java/org/apache/doris/common/util/PropertyAnalyzer.java [hunk 2]
```diff
@@ -130,21 +139,48 @@ public static DataProperty analyzeDataProperty(Map<String, String> properties, D
...
         }
 
         properties.remove(PROPERTIES_STORAGE_MEDIUM);
         properties.remove(PROPERTIES_STORAGE_COLDOWN_TIME);
+        properties.remove(PROPERTIES_STORAGE_COLD_MEDIUM);
+        properties.remove(PROPERTIES_REMOTE_STORAGE);
+
+        if ((hasColdMedium && !hasRemoteStorage) || (!hasColdMedium && hasRemoteStorage)) {
+            throw new AnalysisException("Invalid data property, " +
```
Comment (pengxiangyu) [inline fe/fe-core/src/main/java/org/apache/doris/common/util/PropertyAnalyzer.java:165]: SSD->HDD的情况下不需要remote storage

**Review by pengxiangyu — CHANGES_REQUESTED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/common/util/PropertyAnalyzer.java [hunk 3]
```diff
@@ -130,22 +139,50 @@ public static DataProperty analyzeDataProperty(Map<String, String> properties, D
                 hasCooldown = true;
                 DateLiteral dateLiteral = new DateLiteral(value, Type.DATETIME);
                 coolDownTimeStamp = dateLiteral.unixTimestamp(TimeUtils.getTimeZone());
+            } else if (!hasColdMedium && key.equalsIgnoreCase(PROPERTIES_STORAGE_COLD_MEDIUM)) {
+                hasColdMedium = true;
```
Comment (pengxiangyu) [inline fe/fe-core/src/main/java/org/apache/doris/common/util/PropertyAnalyzer.java:143]: coldMedium有可能是HDD。

File: fe/fe-core/src/main/java/org/apache/doris/catalog/RemoteStorageMgr.java [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+        if (info == null) {
+            throw new AnalysisException("Unknown remote storage name: " + storageName);
+        }
+        return info.getRemoteStorageProperty();
+    }
+
+    public void addRemoteStorage(AddRemoteStorageClause clause) throws DdlException {
+        lock.lock();
+        try {
+            String storageName = clause.getStorageName();
```
Comment (pengxiangyu) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/RemoteStorageMgr.java:76]: storage_name是key，不要引入空格，这里和下面的modify需要做一下trim。

**Review by pengxiangyu — CHANGES_REQUESTED** [review]

**Review by pengxiangyu — APPROVED** [review]
LGTM

**Comment by yiguolei** [comment]
Backup Restore has also defined similar remote storage logic. And Cold Hot defines another one. Could we unify these concepts?

**Comment by qidaye** [comment]
This PR is closed, please refer to #8808

**Pull request closed without merging**